url2a908001cd_malictus_klang_tgz-pJ8-malictus_klang_test_PrimitiveTestJ8/src/malictus/klang/KlangUtil.java

```diff
@@ -0,0 +1,720 @@
+				if (x != (rafDest.getFilePointer() - start)) {
+					throw new IOException(KlangConstants.ERROR_GENERAL_FILE);
+				}
+				rafDest.seek(start + moveAmt);
+				rafDest.write(buf);
+			}
+			buf = new byte[KlangConstants.BUFFER_SIZE];
+			rafSource.seek(0);
+			rafDest.seek(start);
+			while ((rafSource.getFilePointer() + KlangConstants.BUFFER_SIZE) <= rafSource.length()) {
+				KlangProgressWindow.progVal = KlangProgressWindow.PROG_MAX - (int)(((float)(rafSource.length() - rafSource.getFilePointer()) / (float)(rafSource.length() - start)) * KlangProgressWindow.PROG_MAX);
+				int x = rafSource.read(buf);
+				if (x != buf.length) {
+					throw new IOException(KlangConstants.ERROR_GENERAL_FILE);
+				}
+				rafDest.write(buf);
+			}
+			if (rafSource.getFilePointer() < rafSource.length()) {
+				buf = new byte[(int)(rafSource.length() - rafSource.getFilePointer())];
+				int x = rafSource.read(buf);
+				if (x != buf.length) {
+					throw new IOException(KlangConstants.ERROR_GENERAL_FILE);
+				}
+				rafDest.write(buf);
+			}
+			rafDest.close();
+			rafSource.close();
+		} catch (IOException err) {
+			rafDest.close();
+			rafSource.close();
+			throw err;
+		} 
+    }
+    
+	/**
+	 * Copy a file.
+	 * @param source the source file to copy
+	 * @param dest the destination file; if file already exists, it will be overwritten
+	 * @throws IOException if read/write error occurs
+	 */
+    public static void copyFile(File source, File dest) throws IOException {
+    	if (dest.getPath().equals(source.getPath())) {
+			throw new IOException(KlangConstants.ERROR_IN_OUT_IDENTICAL);
+		}
+    	if (dest.exists()) {
+			dest.delete();
+		}
+		dest.createNewFile();
+		FileInputStream fin = new FileInputStream(source);
+		FileOutputStream fos = new FileOutputStream(dest);
+		try {
+			byte[] buffer = new byte[KlangConstants.BUFFER_SIZE];
+	        int len;
+	        while ((len = fin.read(buffer)) > 0) {
+	            fos.write(buffer, 0, len);
+	        }
+	        fin.close();
+	        fos.close();
+		} catch (IOException err) {
+			fin.close();
+			fos.close();
+			throw err;
+		}
+	}
+	
+    /**
+	 * Write data from a source file to a destination file. Any existing data in the destination file will be overwritten.
+	 * @param source the source file
+	 * @param dest the destination file. If this file doesn't exist, it will be created first.
+	 * @param sourceStart start position in the source file for the data to be copied
+	 * @param sourceEnd end position in the source file for the data to be copied
+	 * @throws IOException if read/write errors occur
+	 */
+	public static void writeToFile(File source, File dest, long sourceStart, long sourceEnd) throws IOException {
+		if (dest.getPath().equals(source.getPath())) {
+			throw new IOException(KlangConstants.ERROR_IN_OUT_IDENTICAL);
+		}
+		if (dest.exists()) {
+			dest.delete();
+    	}
+		dest.createNewFile();
+		RandomAccessFile fin = new RandomAccessFile(source, "r");
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fin*. The resource is referenced at lines 563, 565, 567, 568, 569. The resource is closed at lines 575, 578. There are other execution paths that do not contain closure statements, for example, when RandomAccessFile constructor throws an exception. Ensure *fin* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).